fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java

```diff
@@ -124,12 +125,13 @@ public Response runReport(@PathParam("reportName") @Parameter(description = "rep
             parameterTypeValue = "report";
             String reportType = this.readExtraDataAndReportingService.getReportType(reportName, isSelfServiceUserReport);
             ReportingProcessService reportingProcessService = this.reportingProcessServiceProvider.findReportingProcessService(reportType);
-            if (reportingProcessService != null) {
-                return reportingProcessService.processRequest(reportName, queryParams);
+            if (reportingProcessService == null) {
+                throw new PlatformServiceUnavailableException("err.msg.report.service.implementation.missing",
+                        ReportingProcessServiceProvider.SERVICE_MISSING + reportType, reportType);
```

Comment:
Unfortunately based on my read of the code, I don't think this is quite right:

After this change, we are bailing out if we can't find a reporting process service for the report type. 
Previously, if we couldn't find one, we would continue to execute the "hardcoded" reporting process service in the same function below. 

I think continuing execution is the expected behaviour: for any report that is not of type Pentaho or SMS, the report_type is set to Table. But there is no ReportingProcessService with type Table, so therefore the logic in this function below is used to simply prepare the CSV / PDF file. 

In other words: if we check this in, as far as I can see it would break any report that doesn't have a specific report handler configured. 

Now having said that, a good idea would be to create a "Table" type report handler and move the code from this function to that report handler. And after that we could assume that there's a handler for each type of report, and bail out if we can't find one.

Reply:
Oh... wow, very well spotted. I'll have a go at the re-factoring you're suggesting, that would make a lot of sense, yes.

But before I get to that, I'm struggling with more fundamental classloading issues... urgh!! More soon on JIRA, hopefully.

Reply:
@ptuomola I'm finally (7 weeks later, phew!) picking this up, and I'm struggling to understand how the `reportType: SMS` is intended to work... @vidakovic do you understand that? Can those be run on the `/runreports` API? I think what I'll do is just write a `DefaultReportingProcessService` with the current code to which this will delegate to for both `Table` and `SMS`.. I'll therefore make the `type` on `org.apache.fineract.infrastructure.report.annotation.ReportService` a multivalue. Does that sounds reasonable to you guys?

Reply:
Also not sure how this SMS thing is supposed to work... will have a look at this later today.

Reply:
Actually I guess that one would then more appropriately be our `DatatableReportingProcessService` (in `infrastructure.dataqueries.service`) instead of a `DefaultReportingProcessService` in  `infrastructure.report.service.ReportingProcessService`.... watch for an update on this PR! :smile_cat: